Report better errors when multiple invalid constraints are combined

333fred opened the pull request:
Instead of reporting that a constraint must come first, if there are multiple constraints that must come first, we now report that these constraints cannot be combined. We also suppress these errors if a previous error about inherited constraints was reported. Fixes https://github.com/dotnet/roslyn/issues/45141.

xparadoxical commented:
> if there are multiple constraints that must come first, we now report that these constraints cannot be combined.

does it mean that `var` and `nameof` won't be allowed to be in the same line?

Youssef1313 commented on src/Compilers/CSharp/Portable/CSharpResources.resx:
```diff
@@ -6262,4 +6262,19 @@ To remove the warning, you can use /reference instead (set the Embed Interop Typ
   <data name="ERR_CopyConstructorMustInvokeBaseCopyConstructor" xml:space="preserve">
     <value>A copy constructor in a record must call a copy constructor of the base, or a parameterless object constructor if the record inherits from object.</value>
   </data>
+  <data name="IDS_Class" xml:space="preserve">
+    <value>class</value>
+  </data>
+  <data name="IDS_Struct" xml:space="preserve">
+    <value>struct</value>
+  </data>
+  <data name="IDS_Unmanaged" xml:space="preserve">
+    <value>unmanaged</value>
+  </data>
+  <data name="IDS_Notnull" xml:space="preserve">
+    <value>notnull</value>
+  </data>
```
Those might get localized while they shouldn't?

333fred commented:
> 
> 
> > if there are multiple constraints that must come first, we now report that these constraints cannot be combined.
> 
> does it mean that `var` and `nameof` won't be allowed to be in the same line?

@xparadoxical sorry, I copied the wrong URL in the description. This PR is unrelated to that bug.

333fred commented:
@dotnet/roslyn-compiler for reviews on this small diagnostic message cleanup

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_Constraints.cs:
```diff
@@ -286,6 +298,18 @@ static void reportOverrideWithConstraints(ref bool reportedOverrideWithConstrain
+            {
+                if (firstConstraintString is MessageID firstConstraint)
+                {
+                    diagnostics.Add(ErrorCode.ERR_CannotCombineTypeConstraints, constraintLocation, currentConstraintString.Localize(), firstConstraint.Localize());
```
>currentConstraintString.Localize(), firstConstraint.Localize() [](start = 100, length = 62)

Constraint keywords shouldn't be localized.

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_Constraints.cs:
```diff
@@ -126,6 +126,7 @@ internal partial class Binder
             Debug.Assert(!InExecutableBinder); // Cannot eagerly report diagnostics handled by LazyMissingNonNullTypesContextDiagnosticInfo 
             bool hasTypeLikeConstraint = false;
             bool reportedOverrideWithConstraints = false;
+            MessageID? firstTypeConstraintString = null;
```
>MessageID? firstTypeConstraintString = null; [](start = 12, length = 44)

The implementation feels unnecessary complicated. I suggest that we simply report a message that says: "The 'class', 'struct', 'unmanaged' and 'notnull' constraints cannot be repeated or combined with each other and should be specified first." All that complicated tracking system will be unnecessary and we can easily ensure that this error is reported only once.

AlekseyTs commented on src/Compilers/CSharp/Portable/Errors/MessageID.cs:
```diff
@@ -206,6 +206,10 @@ internal enum MessageID
         IDS_FeatureInitOnlySetters = MessageBase + 12781,
         IDS_FeatureRecords = MessageBase + 12782,
         IDS_FeatureNullPointerConstantPattern = MessageBase + 12783,
+        IDS_Class = MessageBase + 12784,
```
>IDS_Class = MessageBase + 12784, [](start = 8, length = 32)

Addition of these strings doesn't feel right

AlekseyTs commented:
Done with review pass (iteration 2), tests are not looked at

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_Constraints.cs:
```diff
@@ -137,14 +138,18 @@ internal partial class Binder
-                            diagnostics.Add(ErrorCode.ERR_RefValBoundMustBeFirst, syntax.GetFirstToken().GetLocation());
+                            if (!reportedOverrideWithConstraints)
+                            {
+                                reportNotFirstOrCannotCombine(diagnostics, syntax.GetFirstToken().GetLocation(), ErrorCode.ERR_RefValBoundMustBeFirst, MessageID.IDS_Class, firstTypeConstraintString);
```
>reportNotFirstOrCannotCombine(diagnostics, syntax.GetFirstToken().GetLocation(), ErrorCode.ERR_RefValBoundMustBeFirst, MessageID.IDS_Class, firstTypeConstraintString); [](start = 32, length = 167)

What is going to be reported for a scenario when a ```class``` constraint is specified twice?

333fred replied:
Simplified and added a test case.

---
In reply to: [441187653](https://github.com/dotnet/roslyn/pull/45147#discussion_r441187653) [](ancestors = 441187653)

333fred commented:
@AlekseyTs addressed feedback.

---
In reply to: [645028784](https://github.com/dotnet/roslyn/pull/45147#issuecomment-645028784) [](ancestors = 645028784)

AlekseyTs commented:
@333fred It looks like there are legitimate test failures.

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/Semantics/GenericConstraintsTests.cs:
```diff
@@ -1827,9 +1827,9 @@ public void UnmanagedConstraint_Compilation_ReferenceType()
 
             c.VerifyDiagnostics(
-                // (1,39): error CS8380: The 'unmanaged' constraint must come before any other constraints
+                // (1,39): error CS8869: The 'unmanaged' constraint cannot be combined with the 'class' constraint
```
>error CS8869: The 'unmanaged' constraint cannot be combined with the 'class' constraint [](start = 27, length = 87)

It looks like the error number and the wording no longer matches the latest revision. Consider updating. in other tests too.

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/Semantics/GenericConstraintsTests.cs:
```diff
@@ -4152,5 +4152,112 @@ public static unsafe void Main()
+                Diagnostic(ErrorCode.ERR_TypeConstraintsMustBeUniqueAndFirst, "notnull").WithLocation(2, 51),
+                // (4,37): error CS0449: The 'class', 'struct', 'unmanaged', and 'notnull' constraints cannot be combined or duplicated, and must be specified first in the constraints list.
+                //     void M1<T2>() where T2 : class, struct, unmanaged, notnull {}
+                Diagnostic(ErrorCode.ERR_TypeConstraintsMustBeUniqueAndFirst, "struct").WithLocation(4, 37),
```
>struct [](start = 79, length = 6)

Should this be "class"?

AlekseyTs commented:
Done with review pass (iteration 3)

333fred commented:
@AlekseyTs addressed feedback.

---
In reply to: [645772487](https://github.com/dotnet/roslyn/pull/45147#issuecomment-645772487) [](ancestors = 645772487)
